networkx/algorithms/alphacore.py

```diff
@@ -0,0 +1,364 @@
+        Higher values (>0.1) speed up execution but lower ranking resolution.
+        Lower values (<0.1) provide finer ranking but increase runtime.
+    start_epsi : float, optional (default=1)
+        The epsilon to start with. Removes all nodes with depth>epsilon initially.
+        Higher values (>1.0) remove more nodes early, emphasizing denser cores.
+        Lower values (<1.0) refine ranking progressively.
+    expo_decay : bool, optional (default=False)
+        If True, dynamically reduces the step size to have high cores with few nodes.
+        Recommended when ranking precision is more important than speed.
+
+    Returns
+    -------
+    dict
+        A dictionary mapping node IDs to dictionaries containing ``alpha`` and
+        ``batch_id`` values.
+
+    Notes
+    -----
+    AlphaCore ranks nodes when each node carries several numeric attributes
+    in addition to its position in the graph.  At every outer iteration the
+    algorithm removes all nodes whose Mahalanobis depth is at least the
+    current threshold *epsi*.  Every node removed together is assigned the
+    same alpha value, defined by
+
+        alpha = 1 − epsi
+
+    Larger alpha means membership in a denser, more central core.  The idea
+    is analogous to the core number returned by the core_number function,
+    but it is based on multivariate attributes rather than degree alone.
+
+    Typical applications include social networks with activity or influence
+    scores, financial or blockchain graphs with liquidity or risk measures,
+    and infrastructure networks where nodes have capacity or reliability
+    values.
+
+    The function returns alpha values in the interval 0 to 1 (higher means
+    more central) and a batch_id counter that starts at 0.
+
+    See Also
+    --------
+    core_number, k_core, degree_centrality
+
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>> G = nx.DiGraph()
+    >>> G.add_nodes_from([(0, {"f1": 0.2}), (1, {"f1": 0.7})])
+    >>> G.add_edge(0, 1, value=3)
+    >>> result = nx.alpha_core(G, features=["f1"])
+    >>> result  # doctest: +SKIP
+    {0: {'alpha': 0.2, 'batch_id': 3}, 1: {'alpha': 0.7, 'batch_id': 9}}
+
+    References
+    ----------
+    .. [1] F. Victor, C.G. Akcora, Y.R. Gel, M. Kantarcioglu.
+       "AlphaCore: Data Depth based Core Decomposition."
+       In: Proceedings of the 27th ACM SIGKDD Conf. on Knowledge Discovery and Data Mining (KDD '21).
+    """
+    import numpy as np
+
+    # Make a copy of the graph to avoid modifying the input
+    graph = G.copy()
+
+    # Handle empty graph case early
+    if graph.number_of_nodes() == 0:
+        return {}
+
+    if features is None:
+        features = ["all"]
+
+    # Extract numerical node features
+    data = _extract_features(graph, features)
+
+    # Compute covariance matrix for depth calculations
+    matrix = _get_feature_matrix(data)
+    cov = np.cov(matrix.T)
```

Comment:
The covariance matrix is computed once for the "full" graph (i.e. before any node removals) here... (continue to comment on L138)